cpp/src/sort/rank.cu

```diff
@@ -0,0 +1,299 @@
+      sorted_order_view,
+      rank_data,
+      [] __device__(auto rank_count1, auto rank_count2) {
+        return MinCount{
+            std::min(thrust::get<0>(rank_count1), thrust::get<0>(rank_count2)),
+            thrust::get<1>(rank_count1) + thrust::get<1>(rank_count2)};
+      },
+      [] __device__(MinCount minrank_count) {  // min+(count-1)/2
+        return static_cast<double>(thrust::get<0>(minrank_count)) +
+              (static_cast<double>(thrust::get<1>(minrank_count)) - 1) / 2.0;
+      },
+      stream);
+}
+
+} // anonymous namespace
+
+std::unique_ptr<column> rank(column_view const &input, rank_method method,
+                            order column_order, include_nulls _include_nulls,
+                            null_order null_precedence,
+                            bool percentage,
+                            rmm::mr::device_memory_resource *mr,
+                            cudaStream_t stream = 0) {
+  auto const input_size = input.size();
+
+  std::unique_ptr<column> rank_column = [&] {
+    // na_option=keep assign NA to NA values
+    if (_include_nulls == include_nulls::NO)
+      return make_numeric_column(data_type(FLOAT64), input_size, copy_bitmask(input, stream, mr), input.null_count(), stream, mr);
+    else
+      return make_numeric_column(data_type(FLOAT64), input_size, mask_state::UNALLOCATED, stream, mr);
+  }();
+  auto rank_mutable_view = rank_column->mutable_view();
+  auto rank_data = rank_mutable_view.data<double>();
+
+  std::unique_ptr<column> sorted_order = (method == rank_method::FIRST)
+          ? detail::stable_sorted_order(table_view{{input}}, {column_order}, {null_precedence}, mr, stream)
+          : detail::sorted_order       (table_view{{input}}, {column_order}, {null_precedence}, mr, stream);
+  column_view sorted_order_view = sorted_order->view();
+
+  // All equal values have same rank and rank always increases by 1 between groups
+  // acts as key for min, max, average to denote equal value groups
+  rmm::device_vector<size_type> dense_rank_sorted = [&] {
+    if (method != rank_method::FIRST)
+      return sorted_dense_rank(input, sorted_order_view, stream);
+    else
+      return rmm::device_vector<size_type>();
+  }();
+
+  switch (method) {
+    case rank_method::FIRST:
+      rank_first  (sorted_order_view, rank_mutable_view, stream);
+      break;
+    case rank_method::DENSE:
+      rank_dense  (dense_rank_sorted, sorted_order_view, rank_mutable_view, stream);
+      break;
+    case rank_method::MIN:
+      rank_min    (dense_rank_sorted, sorted_order_view, rank_mutable_view, stream);
+      break;
+    case rank_method::MAX:
+      rank_max    (dense_rank_sorted, sorted_order_view, rank_mutable_view, stream);
+      break;
+    case rank_method::AVERAGE:
+      rank_average(dense_rank_sorted, sorted_order_view, rank_mutable_view, stream);
+      break;
+    default:
+      CUDF_FAIL("Unexpected rank_method for rank()");
+  }
```

Comment:
This is a really nice clean up!! Awesome work! Not sure if you agree, but might look even better like this:
```cpp
  switch (method) {
    case rank_method::FIRST:   rank_first   (sorted_order_view, rank_mutable_view, stream);                    break;
    case rank_method::DENSE:   rank_dense   (dense_rank_sorted, sorted_order_view, rank_mutable_view, stream); break;
    case rank_method::MIN:     rank_min     (dense_rank_sorted, sorted_order_view, rank_mutable_view, stream); break;
    case rank_method::MAX:     rank_max     (dense_rank_sorted, sorted_order_view, rank_mutable_view, stream); break;
    case rank_method::AVERAGE: rank_average (dense_rank_sorted, sorted_order_view, rank_mutable_view, stream); break;
    default: CUDF_FAIL("Unexpected rank_method for rank()");
  }
```

Reply:
This looks great. is it possible to enforce this style using .clang-format?

Reply:
Not only is it not possible in clang-format, i think clang-format will unformat this. So when we add clang-format in 0.14, we would have to wrap this in preprocessors to protect this. 

This is my single biggest complaint with clang-format. In certain cases where alignment can really make code look beautiful and much easier to read, clang-format will destroy it.